projects/storefrontlib/src/cms-components/checkout/services/checkout-step.service.ts

```diff
@@ -48,47 +48,110 @@ export class CheckoutStepService {
-      activatedRoute
-    );
+    const previousUrl = this.getPreviousCheckoutStepUrl(activatedRoute);
     this.routingService.go(previousUrl === null ? 'cart' : previousUrl);
   }
 
   next(activatedRoute: ActivatedRoute): void {
-    const nextUrl = this.checkoutConfigService.getNextCheckoutStepUrl(
-      activatedRoute
-    );
+    const nextUrl = this.getNextCheckoutStepUrl(activatedRoute);
     this.routingService.go(nextUrl);
   }
 
   getBackBntText(activatedRoute: ActivatedRoute): string {
-    if (
-      this.checkoutConfigService.getPreviousCheckoutStepUrl(activatedRoute) ===
-      null
-    ) {
+    if (this.getPreviousCheckoutStepUrl(activatedRoute) === null) {
       return 'checkout.backToCart';
     }
-
     return 'common.back';
   }
 
   resetSteps(): void {
-    this.intialSteps = this.checkoutConfigService.steps
+    this.allSteps = this.checkoutConfigService.steps
       .filter((step) => !step.disabled)
       .map((x) => Object.assign({}, x));
-    this.steps$.next(this.intialSteps);
+    this.steps$.next(this.allSteps);
   }
 
   disableEnableStep(
     currentStepType: CheckoutStepType,
     disabled: boolean
   ): void {
-    const currentStep = this.intialSteps.find((step) =>
+    const currentStep = this.allSteps.find((step) =>
       step.type.includes(currentStepType)
     );
     if (currentStep) {
       currentStep.disabled = disabled;
-      this.steps$.next(this.intialSteps.filter((step) => !step.disabled));
+      this.steps$.next(this.allSteps.filter((step) => !step.disabled));
+    }
+  }
+
+  getNextCheckoutStepUrl(activatedRoute: ActivatedRoute): string {
+    const stepIndex = this.getCurrentStepIndex(activatedRoute);
+
+    if (stepIndex >= 0) {
+      let i = 1;
+      while (
+        this.allSteps[stepIndex + i] &&
+        this.allSteps[stepIndex + i].disabled
+      ) {
+        i++;
+      }
+      const nextStep = this.allSteps[stepIndex + i];
+      if (nextStep) {
+        return this.getStepUrlFromStepRoute(nextStep.routeName);
+      }
     }
+    return null;
+  }
+
+  getPreviousCheckoutStepUrl(activatedRoute: ActivatedRoute): string {
+    const stepIndex = this.getCurrentStepIndex(activatedRoute);
+
+    if (stepIndex >= 0) {
+      let i = 1;
+      while (
+        this.allSteps[stepIndex - i] &&
+        this.allSteps[stepIndex - i].disabled
+      ) {
+        i++;
+      }
+      const previousStep = this.allSteps[stepIndex - i];
+      if (previousStep) {
+        return this.getStepUrlFromStepRoute(previousStep.routeName);
+      }
+    }
+    return null;
+  }
+
+  getCurrentStepIndex(activatedRoute: ActivatedRoute): number | null {
+    const currentStepUrl: string = this.getStepUrlFromActivatedRoute(
+      activatedRoute
+    );
+
+    let stepIndex: number;
+    let index = 0;
+    for (const step of this.allSteps) {
+      if (
+        currentStepUrl === `/${this.getStepUrlFromStepRoute(step.routeName)}`
+      ) {
+        stepIndex = index;
```

Comment:
We could `break;` when we find it.

Or perhaps even better, would the `findIndex` array function fit the need here?  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex

Reply:
good catch, I will use `findIndex`.